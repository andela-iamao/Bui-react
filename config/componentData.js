module.exports = [{"name":"Button","description":"Button Component","props":{"children":{"type":{"name":"node"},"required":true,"description":"Child elements for the button component"},"onClick":{"type":{"name":"func"},"required":false,"description":""},"type":{"type":{"name":"enum","value":[{"value":"'primary'","computed":false},{"value":"'secondary'","computed":false},{"value":"'danger'","computed":false}]},"required":false,"description":""}},"code":"import React from 'react';\nimport PropTypes from 'prop-types';\n\n\nconst colorScheme = {\n    primary: '#236167',\n    secondary: '#000000',\n    danger: '#AA3939'\n}\n\n/**\n * Button Component\n */\nfunction Button({ children, ...props}) {\n    return (\n        <button {...props} style={style(props.type)}>\n            {children}\n        </button>\n    );\n}\n\nconst style = (type) => ({\n    color: '#FFFFFF',\n    backgroundColor: colorScheme[type],\n    padding: 10,\n    width: 120,\n    border: 'none'\n});\n\nButton.propTypes = {\n    /**\n     * Child elements for the button component\n     */\n    children: PropTypes.node.isRequired,\n    onClick: PropTypes.func,\n    type: PropTypes.oneOf(['primary', 'secondary', 'danger'])\n}\n\nexport default Button;\n","examples":[{"name":"DangerButton","description":"Danger Button component","code":"import React from 'react';\nimport Button from 'bui-react/Button';\n\n/**\n * Danger Button component\n */\nexport default () => {\n    return (\n        <Button type=\"danger\">Danger Button!</Button>\n    );\n}\n"},{"name":"PrimaryButton","description":"Primary Button component","code":"import React from 'react';\nimport Button from 'bui-react/Button';\n\n/**\n * Primary Button component\n */\nexport default () => {\n    return (\n        <Button type=\"primary\">Primary Button!</Button>\n    );\n}\n"},{"name":"SecondaryButton","description":"Secondary Button component","code":"import React from 'react';\nimport Button from 'bui-react/Button';\n\n/**\n * Secondary Button component\n */\nexport default () => {\n    return (\n        <Button type=\"secondary\">Secondary Button!</Button>\n    );\n}\n"}]}]