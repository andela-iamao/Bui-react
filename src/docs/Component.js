import React from 'react';
import PropTypes from 'prop-types';

import Props from './Props';
import Example from './Example';


const Component = ({component}) => {
    const { name, description, props, examples } = component;

    return (
        <div id="bui-react-docs-component-container">
            <h1 id="bui-react-docs-component-name">{name}</h1>
            <p id="bui-react-docs-component-description">{description}</p>
            <br /><br />
            <h3 id="bui-react-docs-component-example-header">Example{examples.length > 1 && 's'}</h3>
            {
                examples.length > 0 ?
                    examples.map((example, index) => (
                        <Example
                            example={example}
                            key={example.code}
                            componentName={name}
                            last={index === examples.length - 1}
                        />
                    ))
                    :
                    "No examples yet"
            }
            <br /><br />
            <h3 id="bui-react-docs-component-props-header">Props</h3>
            {
                props ?
                    <Props props={props} />
                    :
                    "This component accepts no props"
            }
        </div>
    )
};


Component.propTypes = {
    component: PropTypes.object.isRequired
};

export default Component;
