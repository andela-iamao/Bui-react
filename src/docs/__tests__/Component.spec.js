import React from 'react';
import renderer from 'react-test-renderer';

import { mount, shallow } from 'enzyme';

import Component from '../Component';
import componentData from '../../../config/jest/componentDataMock';

const DATA = componentData[0];


describe('DOCS: Component', () => {

    describe('Snapshot Test', () => {
        test('renders component with examples alone', () => {
            const data = { ...componentData[0]};
            data.props = null;
            const tree = renderer.create(<Component component={data} />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        test('renders component with props alone', () => {
            const data = { ...componentData[0]};
            data.examples = [];
            const tree = renderer.create(<Component component={data} />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        test('renders component with props and examples', () => {
            const tree = renderer.create(<Component component={DATA} />).toJSON();
            expect(tree).toMatchSnapshot();
        });

        test('renders component without props and examples', () => {
            const data = { ...componentData[0]};
            data.examples = [];
            data.props = null;
            const tree = renderer.create(<Component component={data} />).toJSON();
            expect(tree).toMatchSnapshot();
        });
    });

    describe('Integration Test', () => {
        test('renders and mount the component', () => {
            const wrapper = mount(<Component component={DATA} />)
            expect(wrapper.find({ id: 'bui-react-docs-component-example-container' })).toHaveLength(1);
            expect(wrapper.find({ id: 'bui-react-docs-component-props-container' })).toHaveLength(1);
        });

        test('renders without examples if no examples in metadata', () => {
            const data = { ...DATA};
            data.examples = [];
            const wrapper = mount(<Component component={data} />)
            expect(wrapper.find({ id: 'bui-react-docs-component-example-container' })).toHaveLength(0);
            expect(wrapper.find({ id: 'bui-react-docs-component-props-container' })).toHaveLength(1);
        });

        test('renders without props if no props in metadata', () => {
            const data = { ...DATA};
            data.props = null;
            const wrapper = mount(<Component component={data} />);
            expect(wrapper.find({ id: 'bui-react-docs-component-example-container' })).toHaveLength(1);
            expect(wrapper.find({ id: 'bui-react-docs-component-props-container' })).toHaveLength(0);
        });
    });

    describe('Unit Test', () => {
        test('renders and displays basic metadata about component', () => {
            const wrapper = shallow(<Component component={DATA} />);
            expect(wrapper.find({ id: 'bui-react-docs-component-name' }).text()).toEqual(componentData[0].name);
            expect(wrapper.find({ id: 'bui-react-docs-component-description' }).text()).toEqual(componentData[0].description);
        });

        test('should render a plural of "Example" when multiple examples are present', () => {
          const data  = { ...DATA, examples: [...DATA.examples, DATA.examples[0]] };
          const wrapper = shallow(<Component component={data} />);
          expect(wrapper.find('#bui-react-docs-component-example-header').text()).toEqual('Examples');
        });

    });

});

