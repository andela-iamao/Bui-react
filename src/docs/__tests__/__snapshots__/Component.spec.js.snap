// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOCS: Component Snapshot Test renders component with examples alone 1`] = `
<div
  id="bui-react-docs-component-container"
>
  <h1
    id="bui-react-docs-component-name"
  >
    Button
  </h1>
  <p
    id="bui-react-docs-component-description"
  >
    Button Component
  </p>
  <br />
  <br />
  <h3
    id="bui-react-docs-component-example-header"
  >
    Example
  </h3>
  <div
    id="bui-react-docs-component-example-container"
  >
    <h4
      id="bui-react-docs-component-example-description"
    >
      Default Button component
    </h4>
    <button>
      Hello!
    </button>
    <p
      id="bui-react-docs-component-example-code"
    >
      <span
        className="link"
        id="bui-react-docs-component-example-code-toggle-btn"
        onClick={[Function]}
      >
        show
         Code
      </span>
    </p>
  </div>
  <br />
  <br />
  <h3
    id="bui-react-docs-component-props-header"
  >
    Props
  </h3>
  This component accepts no props
</div>
`;

exports[`DOCS: Component Snapshot Test renders component with props alone 1`] = `
<div
  id="bui-react-docs-component-container"
>
  <h1
    id="bui-react-docs-component-name"
  >
    Button
  </h1>
  <p
    id="bui-react-docs-component-description"
  >
    Button Component
  </p>
  <br />
  <br />
  <h3
    id="bui-react-docs-component-example-header"
  >
    Example
  </h3>
  No examples yet
  <br />
  <br />
  <h3
    id="bui-react-docs-component-props-header"
  >
    Props
  </h3>
  <div
    id="bui-react-docs-component-props-container"
  >
    <table
      className="props"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Description
          </th>
          <th>
            Type
          </th>
          <th>
            Default
          </th>
          <th>
            Required
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            children
          </td>
          <td>
            Child elements for the button component
          </td>
          <td>
            node
          </td>
          <td />
          <td>
            Nope
          </td>
        </tr>
        <tr>
          <td>
            onClick
          </td>
          <td>
            
          </td>
          <td>
            func
          </td>
          <td />
          <td />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`DOCS: Component Snapshot Test renders component with props and examples 1`] = `
<div
  id="bui-react-docs-component-container"
>
  <h1
    id="bui-react-docs-component-name"
  >
    Button
  </h1>
  <p
    id="bui-react-docs-component-description"
  >
    Button Component
  </p>
  <br />
  <br />
  <h3
    id="bui-react-docs-component-example-header"
  >
    Example
  </h3>
  <div
    id="bui-react-docs-component-example-container"
  >
    <h4
      id="bui-react-docs-component-example-description"
    >
      Default Button component
    </h4>
    <button>
      Hello!
    </button>
    <p
      id="bui-react-docs-component-example-code"
    >
      <span
        className="link"
        id="bui-react-docs-component-example-code-toggle-btn"
        onClick={[Function]}
      >
        show
         Code
      </span>
    </p>
  </div>
  <br />
  <br />
  <h3
    id="bui-react-docs-component-props-header"
  >
    Props
  </h3>
  <div
    id="bui-react-docs-component-props-container"
  >
    <table
      className="props"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Description
          </th>
          <th>
            Type
          </th>
          <th>
            Default
          </th>
          <th>
            Required
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            children
          </td>
          <td>
            Child elements for the button component
          </td>
          <td>
            node
          </td>
          <td />
          <td>
            Nope
          </td>
        </tr>
        <tr>
          <td>
            onClick
          </td>
          <td>
            
          </td>
          <td>
            func
          </td>
          <td />
          <td />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`DOCS: Component Snapshot Test renders component without props and examples 1`] = `
<div
  id="bui-react-docs-component-container"
>
  <h1
    id="bui-react-docs-component-name"
  >
    Button
  </h1>
  <p
    id="bui-react-docs-component-description"
  >
    Button Component
  </p>
  <br />
  <br />
  <h3
    id="bui-react-docs-component-example-header"
  >
    Example
  </h3>
  No examples yet
  <br />
  <br />
  <h3
    id="bui-react-docs-component-props-header"
  >
    Props
  </h3>
  This component accepts no props
</div>
`;
